In this project , we are predicting customer churn for a telecom company.Churn is when a customer stops using a service. A telecom company has many customers, and every month some of them decide to terminate their contracts and switch to another provider. Losing customers is expensive, so the company wants to identify the people who are likely to leave before they actually do.
The idea is simple: for every customer, we compute a score between 0 and 1 - the probability that this customer is about to churn. Then the company can target the customers with high scores and try to retain them, for example by sending an email with discounts or special promotions.

To predict this score we use machine learning. The task is binary classification: we look at one customer and predict one of two possible outcomes. For the i-th customer, we want our model g to approximate the target y:

g(xi) ≈ yi

In the formula, yi is the model's prediction and belongs to {0,1}, with 0 being the negative value or no churning, and 1 the positive value or churning. The output corresponds to the likelihood of churning.

In brief, the main idea behind this project is to build a model with historical data from customers and assign a score of the likelihood of churning.

## Data Preparation 
* Download data and read it with Pandas 
* Look at the data and understand 
* Make columns names and vlaues look uniform 
* Check if all columns read correctly 
* Check if the churn varialbe needs any preparation

In [ ]:
import pandas as pd # To import pandas library for data manipulation and analysis
import numpy as np # To import numpy library for numerical computations

import matplotlib.pyplot as plt # To import matplotlib library for data visualization
import seaborn as sns # To import seaborn library for statistical data visualization

# To display plots inline in Jupyter Notebook
%matplotlib inline 

Dataset used in this project is donwloded from Kaggle , URL: https://www.kaggle.com/datasets/blastchar/telco-customer-churn

In [ ]:
df = pd.read_csv('WA_Fn-UseC_-Telco-Customer-Churn.csv') # To read the CSV file containing customer churn data into a pandas DataFrame
df

,customerID,gender,SeniorCitizen,Partner,Dependents,tenure,PhoneService,MultipleLines,InternetService,OnlineSecurity,...,DeviceProtection,TechSupport,StreamingTV,StreamingMovies,Contract,PaperlessBilling,PaymentMethod,MonthlyCharges,TotalCharges,Churn
0,7590-VHVEG,Female,0,Yes,No,1,No,No phone service,DSL,No,...,No,No,No,No,Month-to-month,Yes,Electronic check,29.85,29.85,No
1,5575-GNVDE,Male,0,No,No,34,Yes,No,DSL,Yes,...,Yes,No,No,No,One year,No,Mailed check,56.95,1889.5,No
2,3668-QPYBK,Male,0,No,No,2,Yes,No,DSL,Yes,...,No,No,No,No,Month-to-month,Yes,Mailed check,53.85,108.15,Yes
3,7795-CFOCW,Male,0,No,No,45,No,No phone service,DSL,Yes,...,Yes,Yes,No,No,One year,No,Bank transfer (automatic),42.30,1840.75,No
4,9237-HQITU,Female,0,No,No,2,Yes,No,Fiber optic,No,...,No,No,No,No,Month-to-month,Yes,Electronic check,70.70,151.65,Yes
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
7038,6840-RESVB,Male,0,Yes,Yes,24,Yes,Yes,DSL,Yes,...,Yes,Yes,Yes,Yes,One year,Yes,Mailed check,84.80,1990.5,No
7039,2234-XADUH,Female,0,Yes,Yes,72,Yes,Yes,Fiber optic,No,...,Yes,No,Yes,Yes,One year,Yes,Credit card (automatic),103.20,7362.9,No
7040,4801-JZAZL,Female,0,Yes,Yes,11,No,No phone service,DSL,Yes,...,No,No,No,No,Month-to-month,Yes,Electronic check,29.60,346.45,No
7041,8361-LTMKD,Male,1,Yes,No,4,Yes,Yes,Fiber optic,No,...,No,No,No,No,Month-to-month,Yes,Mailed check,74.40,306.6,Yes


In [ ]:
df.head().T # To show the first 5 rows of the DataFrame and transpose it for better readability, which converts rows to columns and columns to rows. This is useful for quickly inspecting the data and understanding its structure.

,0,1,2,3,4
customerID,7590-VHVEG,5575-GNVDE,3668-QPYBK,7795-CFOCW,9237-HQITU
gender,Female,Male,Male,Male,Female
SeniorCitizen,0,0,0,0,0
Partner,Yes,No,No,No,No
Dependents,No,No,No,No,No
tenure,1,34,2,45,2
PhoneService,No,Yes,Yes,No,Yes
MultipleLines,No phone service,No,No,No phone service,No
InternetService,DSL,DSL,DSL,DSL,Fiber optic
OnlineSecurity,No,Yes,Yes,Yes,No


In [ ]:
df.columns = df.columns.str.lower().str.replace(' ', '_') # To convert all column names to lowercase and replace spaces with underscores for consistency and easier access

categorical_columns = list(df.dtypes[df.dtypes == 'str'].index) # To create a list of column names that have data type 'object', which typically indicates categorical variables

for col in categorical_columns: # To iterate over each column name in the list of categorical columns
    df[col] = df[col].str.lower().str.replace(' ', '_') # To convert all string values in the column to lowercase and replace spaces with underscores for consistency and easier analysis
    
df

,customerid,gender,seniorcitizen,partner,dependents,tenure,phoneservice,multiplelines,internetservice,onlinesecurity,...,deviceprotection,techsupport,streamingtv,streamingmovies,contract,paperlessbilling,paymentmethod,monthlycharges,totalcharges,churn
0,7590-vhveg,female,0,yes,no,1,no,no_phone_service,dsl,no,...,no,no,no,no,month-to-month,yes,electronic_check,29.85,29.85,no
1,5575-gnvde,male,0,no,no,34,yes,no,dsl,yes,...,yes,no,no,no,one_year,no,mailed_check,56.95,1889.5,no
2,3668-qpybk,male,0,no,no,2,yes,no,dsl,yes,...,no,no,no,no,month-to-month,yes,mailed_check,53.85,108.15,yes
3,7795-cfocw,male,0,no,no,45,no,no_phone_service,dsl,yes,...,yes,yes,no,no,one_year,no,bank_transfer_(automatic),42.30,1840.75,no
4,9237-hqitu,female,0,no,no,2,yes,no,fiber_optic,no,...,no,no,no,no,month-to-month,yes,electronic_check,70.70,151.65,yes
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
7038,6840-resvb,male,0,yes,yes,24,yes,yes,dsl,yes,...,yes,yes,yes,yes,one_year,yes,mailed_check,84.80,1990.5,no
7039,2234-xaduh,female,0,yes,yes,72,yes,yes,fiber_optic,no,...,yes,no,yes,yes,one_year,yes,credit_card_(automatic),103.20,7362.9,no
7040,4801-jzazl,female,0,yes,yes,11,no,no_phone_service,dsl,yes,...,no,no,no,no,month-to-month,yes,electronic_check,29.60,346.45,no
7041,8361-ltmkd,male,1,yes,no,4,yes,yes,fiber_optic,no,...,no,no,no,no,month-to-month,yes,mailed_check,74.40,306.6,yes


In [ ]:
df.dtypes
# We notice some columns like 'totalcharges' are of object type, which should be numeric. We will convert them to numeric type and handle any errors that arise during the conversion.
# And seniorcitizen column is of int type, we will convert it to object type for consistency with other categorical variables.


customerid              str
gender                  str
seniorcitizen         int64
partner                 str
dependents              str
tenure                int64
phoneservice            str
multiplelines           str
internetservice         str
onlinesecurity          str
onlinebackup            str
deviceprotection        str
techsupport             str
streamingtv             str
streamingmovies         str
contract                str
paperlessbilling        str
paymentmethod           str
monthlycharges      float64
totalcharges            str
churn                   str
dtype: object

In [ ]:
tc = pd.to_numeric(df.totalcharges, errors='coerce') # To convert the 'totalcharges' column to numeric type, coercing any errors (e.g., non-numeric values) to NaN
df.totalcharges = tc # To replace the original 'totalcharges' column with the converted numeric column

df[tc.isnull()][['customerid' , 'totalcharges']] # To display rows where the 'totalcharges' column has NaN values, showing the 'customerid' and 'totalcharges' columns for those rows. This helps identify which customers have missing or invalid total charges.

,customerid,totalcharges
488,4472-lvygi,NaN
753,3115-czmzd,NaN
936,5709-lvoeq,NaN
1082,4367-nuyao,NaN
1340,1371-dwpaz,NaN
3331,7644-omvmy,NaN
3826,3213-vvolg,NaN
4380,2520-sgtta,NaN
5218,2923-arzlg,NaN
6670,4075-wkniu,NaN


In [ ]:
df.totalcharges = df.totalcharges.fillna(0) # To fill any NaN values in the 'totalcharges' column with 0, ensuring that there are no missing values in this numeric column

In [ ]:
df[tc.isnull()][['customerid' , 'totalcharges']] # To display the rows where 'totalcharges' could not be converted to numeric (i.e., where it is NaN), showing the 'customerid' and 'totalcharges' columns for reference


,customerid,totalcharges
488,4472-lvygi,0.0
753,3115-czmzd,0.0
936,5709-lvoeq,0.0
1082,4367-nuyao,0.0
1340,1371-dwpaz,0.0
3331,7644-omvmy,0.0
3826,3213-vvolg,0.0
4380,2520-sgtta,0.0
5218,2923-arzlg,0.0
6670,4075-wkniu,0.0


In [ ]:
df.churn = (df.churn == 'yes').astype(int) # To convert the 'churn' column to a binary format where 'yes' is represented as 1 and 'no' as 0, making it suitable for machine learning models that require numerical input.
df.churn.head()


0    0
1    0
2    1
3    0
4    1
Name: churn, dtype: int64

## Setting up validation framework

Performing split of data to train , validation and test sets with using Scikit-Learn

In [ ]:
from sklearn.model_selection import train_test_split # To import the train_test_split function from scikit-learn for splitting the dataset into training and testing sets

In [ ]:
# To display the documentation for the train_test_split function, which provides information on its parameters and usage. This is useful for understanding how to properly split the dataset into training and testing subsets.
train_test_split?

Signature:
train_test_split(
    *arrays,
    test_size=None,
    train_size=None,
    random_state=None,
    shuffle=True,
    stratify=None,
)
Docstring:
Split arrays or matrices into random train and test subsets.

Quick utility that wraps input validation,
``next(ShuffleSplit().split(X, y))``, and application to input data
into a single call for splitting (and optionally subsampling) data into a
one-liner.

Read more in the :ref:`User Guide <cross_validation>`.

Parameters
----------
*arrays : sequence of array-like of shape (n_samples,) or             (n_samples, n_outputs)
    Indexable data-structures can be arrays, lists, dataframes, scipy
    sparse matrices or pandas dataframes with consistent first dimension.

test_size : float or int, default=None
    If float, should be between 0.0 and 1.0 and represent the proportion
    of the dataset to include in the test split. If int, represents the
    absolute number of test samples. If None, the value is set to the
    complement 

In [ ]:
df_full_train , df_test = train_test_split(df, test_size = 0.2 , random_state = 1) # To split the DataFrame 'df' into training and testing sets, with 20% of the data allocated for testing and a random state of 1 for reproducibility , so wehen execute the code you will get the same result. This function returns four outputs: X_train, X_test, y_train, y_test.

In [ ]:
df_train , df_val = train_test_split(df_full_train, test_size = 0.25 , random_state = 1) # To further split the training set into a smaller training set and a validation set, with 25% of the data allocated for validation. This allows for model evaluation during training.


In [ ]:
len(df_full_train), len(df_test) # To display the number of rows in the training and testing sets, confirming the split was successful and providing insight into the size of each dataset.

(5634, 1409)

In [ ]:
len(df_train), len(df_val), len(df_test) # To display the number of rows in the training, validation, and testing sets, providing insight into the size of each subset after the splits.

(4225, 1409, 1409)

In [ ]:
df_train = df_train.reset_index(drop = True )
df_val = df_val.reset_index(drop = True ) 
df_test = df_test.reset_index(drop = True ) # To reset the index of the training, validation, and testing DataFrames, dropping the old index. This is useful for ensuring that the indices are sequential and start from 0, which can help avoid confusion during analysis and modeling.

In [ ]:
y_train = df_train.churn.values # To extract the target variable 'churn' from the training set and convert it to a NumPy array for model training
y_val = df_val.churn.values # To extract the target variable 'churn' from the validation set and convert it to a NumPy array for model evaluation
y_test = df_test.churn.values # To extract the target variable 'churn' from the testing set and convert it to a NumPy array for final model evaluation

In [ ]:
del df_train['churn'] # To remove the 'churn' column from the training DataFrame, as it has already been extracted into the y_train array and is no longer needed in the feature set.
del df_val['churn'] # To remove the 'churn' column from the validation DataFrame, as it has already been extracted into the y_val array and is no longer needed in the feature set.
del df_test['churn'] # To remove the 'churn' column from the testing DataFrame, as it has already been extracted into the y_test array and is no

## Exploratory Data Analysis 
* Chech Missing Values 
* Look at the target Varialbe 
* Look at numerical and categorical varialbes 

In [ ]:
df_full_train = df_full_train.reset_index(drop = True ) # To reset the index of the full training DataFrame, dropping the old index. This ensures that the indices are sequential and start from 0, which can help avoid confusion during analysis and modeling.
df_full_train

,customerid,gender,seniorcitizen,partner,dependents,tenure,phoneservice,multiplelines,internetservice,onlinesecurity,...,deviceprotection,techsupport,streamingtv,streamingmovies,contract,paperlessbilling,paymentmethod,monthlycharges,totalcharges,churn
0,5442-pptjy,male,0,yes,yes,12,yes,no,no,no_internet_service,...,no_internet_service,no_internet_service,no_internet_service,no_internet_service,two_year,no,mailed_check,19.70,258.35,0
1,6261-rcvns,female,0,no,no,42,yes,no,dsl,yes,...,yes,yes,no,yes,one_year,no,credit_card_(automatic),73.90,3160.55,1
2,2176-osjuv,male,0,yes,no,71,yes,yes,dsl,yes,...,no,yes,no,no,two_year,no,bank_transfer_(automatic),65.15,4681.75,0
3,6161-erdgd,male,0,yes,yes,71,yes,yes,dsl,yes,...,yes,yes,yes,yes,one_year,no,electronic_check,85.45,6300.85,0
4,2364-ufrom,male,0,no,no,30,yes,no,dsl,yes,...,no,yes,yes,no,one_year,no,electronic_check,70.40,2044.75,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
5629,0781-lkxbr,male,1,no,no,9,yes,yes,fiber_optic,no,...,yes,no,yes,yes,month-to-month,yes,electronic_check,100.50,918.60,1
5630,3507-gasnp,male,0,no,yes,60,yes,no,no,no_internet_service,...,no_internet_service,no_internet_service,no_internet_service,no_internet_service,two_year,no,mailed_check,19.95,1189.90,0
5631,8868-wozgu,male,0,no,no,28,yes,yes,fiber_optic,no,...,yes,no,yes,yes,month-to-month,yes,electronic_check,105.70,2979.50,1
5632,1251-krreg,male,0,no,no,2,yes,yes,dsl,no,...,no,no,no,no,month-to-month,yes,mailed_check,54.40,114.10,1


In [ ]:
df_full_train.isnull().sum() # To check for any missing values in the full training DataFrame by summing the number of null entries in each column. We only had missing values in the 'TotalCharges' column which was fixed in the 'data preparation' step. This helps identify if any data cleaning or imputation is needed before modeling.

customerid          0
gender              0
seniorcitizen       0
partner             0
dependents          0
tenure              0
phoneservice        0
multiplelines       0
internetservice     0
onlinesecurity      0
onlinebackup        0
deviceprotection    0
techsupport         0
streamingtv         0
streamingmovies     0
contract            0
paperlessbilling    0
paymentmethod       0
monthlycharges      0
totalcharges        0
churn               0
dtype: int64

In [ ]:
df_full_train.churn.value_counts() # To count the occurrences of each unique value in the 'churn' column of the full training DataFrame. This provides insight into the distribution of the target variable, which is important for understanding class balance and potential model performance issues related to imbalanced datasets.

churn
0    4113
1    1521
Name: count, dtype: int64

In [ ]:
# Let's check the churn rate in the full training dataset to understand the class distribution and potential imbalance in the target variable. This is important for model evaluation and selection of appropriate metrics.
df_full_train.churn.value_counts(normalize = True) # To display the proportion of each class in the 'churn' column of the full training DataFrame, providing insight into the class distribution and potential class imbalance in the target variable.

churn
0    0.730032
1    0.269968
Name: proportion, dtype: float64

In [ ]:
global_churn_rate = df_full_train.churn.mean() # To calculate the mean of the 'churn' column in the full training DataFrame, which represents the average churn rate. This provides a single value summary of the target variable's distribution and can be useful for understanding the overall churn behavior in the dataset.
round(global_churn_rate, 2) # To round the calculated global churn rate to two decimal places for easier interpretation and presentation. This provides a concise summary of the average churn rate in the dataset.

np.float64(0.27)

In [ ]:
df_full_train.dtypes # To display the data types of each column in the full training DataFrame, providing insight into the structure of the dataset and helping identify which columns are categorical, numerical, or require type conversion for modeling.

# Let's check the number of unique values in each column to understand the diversity of data and identify potential categorical variables that may need encoding for machine learning models.
# seniorcitizen columns is of int type, but not numeric, so we will convert it to object type for consistency with other categorical variables.

customerid              str
gender                  str
seniorcitizen         int64
partner                 str
dependents              str
tenure                int64
phoneservice            str
multiplelines           str
internetservice         str
onlinesecurity          str
onlinebackup            str
deviceprotection        str
techsupport             str
streamingtv             str
streamingmovies         str
contract                str
paperlessbilling        str
paymentmethod           str
monthlycharges      float64
totalcharges        float64
churn                 int64
dtype: object

In [ ]:
numerical = [ 'tenure', 'monthlycharges', 'totalcharges' ] # To create a list of column names that are considered numerical features, which can be used for analysis and modeling.

In [ ]:
categorical = ['gender', 'seniorcitizen', 'partner', 'dependents',
               'phoneservice', 'multiplelines', 'internetservice',
       'onlinesecurity', 'onlinebackup', 'deviceprotection', 'techsupport',
       'streamingtv', 'streamingmovies', 'contract', 'paperlessbilling',
       'paymentmethod'] # To create a list of column names that are considered categorical features, which can be used for analysis and modeling.
# Except ' customerid' and 'churn' columns, all other columns are categorical variables. We will convert the 'seniorcitizen' column to object type for consistency with other categorical variables.

In [ ]:
df_full_train[categorical].nunique() # To display the number of unique values in each categorical column of the full training DataFrame, providing insight into the diversity of data and helping identify potential categorical variables that may need encoding for machine learning models.

gender              2
seniorcitizen       2
partner             2
dependents          2
phoneservice        2
multiplelines       3
internetservice     3
onlinesecurity      3
onlinebackup        3
deviceprotection    3
techsupport         3
streamingtv         3
streamingmovies     3
contract            3
paperlessbilling    2
paymentmethod       4
dtype: int64

# Feature Importance : Churn Rate and Risk Ratio

Feature imporance analysis ( part of EDA ) : identifying which feature affects our target variable

* Churn Rate
* Risk Ratio
* Mutual information

To work on Feature importance , we find the difference between global chur rate and grooup churn rate. 

* If global > group : less likely to churn 
* If global < group : more likely to churn
-------------------------------------------------

Risk Ratio : group churn rate / global churn rate 
 * If risk ratio > 1 : More likely to churn
 * If risk ratio < 1 : Less likely to churn


In [ ]:
df_full_train.head()

,customerid,gender,seniorcitizen,partner,dependents,tenure,phoneservice,multiplelines,internetservice,onlinesecurity,...,deviceprotection,techsupport,streamingtv,streamingmovies,contract,paperlessbilling,paymentmethod,monthlycharges,totalcharges,churn
0,5442-pptjy,male,0,yes,yes,12,yes,no,no,no_internet_service,...,no_internet_service,no_internet_service,no_internet_service,no_internet_service,two_year,no,mailed_check,19.70,258.35,0
1,6261-rcvns,female,0,no,no,42,yes,no,dsl,yes,...,yes,yes,no,yes,one_year,no,credit_card_(automatic),73.90,3160.55,1
2,2176-osjuv,male,0,yes,no,71,yes,yes,dsl,yes,...,no,yes,no,no,two_year,no,bank_transfer_(automatic),65.15,4681.75,0
3,6161-erdgd,male,0,yes,yes,71,yes,yes,dsl,yes,...,yes,yes,yes,yes,one_year,no,electronic_check,85.45,6300.85,0
4,2364-ufrom,male,0,no,no,30,yes,no,dsl,yes,...,no,yes,yes,no,one_year,no,electronic_check,70.40,2044.75,0


In [ ]:
churn_female = df_full_train[df_full_train.gender == 'female'].churn.mean() # To calculate the mean churn rate for female customers in the full training DataFrame, providing insight into the churn behavior of this specific demographic group.
churn_female

np.float64(0.27682403433476394)

In [ ]:
churn_male = df_full_train[df_full_train.gender == 'male'].churn.mean() # To calculate the mean churn rate for male customers in the full training DataFrame, providing insight into the churn behavior of this specific demographic group.
churn_male

np.float64(0.2632135306553911)

In [ ]:
global_churn_rate = df_full_train.churn.mean() # To calculate the mean of the 'churn' column in the full training DataFrame, which represents the average churn rate. This provides a single value summary of the target variable's distribution and can be useful for understanding the overall churn behavior in the dataset.
global_churn_rate

np.float64(0.26996805111821087)

We can do the same fo other cetagories , like partner 

In [ ]:
df_full_train.partner.value_counts()

partner
no     2932
yes    2702
Name: count, dtype: int64

In [ ]:
churn_with_partner = df_full_train[df_full_train.partner == 'yes'].churn.mean() # To calculate the mean churn rate for customers who have a partner in the full training DataFrame, providing insight into the churn behavior of this specific demographic group.
churn_with_partner

np.float64(0.20503330866025166)

In [ ]:
churn_without_partner = df_full_train[df_full_train.partner == 'no'].churn.mean() # To calculate the mean churn rate for customers who do not have a partner in the full training DataFrame, providing insight into the churn behavior of this specific demographic group.
churn_without_partner

np.float64(0.3298090040927694)

We noice with / without partner are significantly different from the global churn rate

In [ ]:
print (global_churn_rate - churn_with_partner)
print (global_churn_rate - churn_without_partner)
print(global_churn_rate - churn_female)
print(global_churn_rate - churn_male)

0.06493474245795922
-0.05984095297455855
-0.006855983216553063
0.006754520462819769


It gives us some ideas hat partner feature is more important in predecting churn than gender 

# Risk Ratio

In [ ]:
churn_with_partner / global_churn_rate # To calculate the ratio of the churn rate for customers with a partner to the global churn rate, providing insight into how having a partner may influence churn behavior compared to the overall average churn rate in the dataset.

np.float64(0.7594724924338315)

In [ ]:
churn_without_partner / global_churn_rate # To calculate the ratio of the churn rate for customers without a partner to the global churn rate, providing insight into how not having a partner may influence churn behavior compared to the overall average churn rate in the dataset.

np.float64(1.2216593879412643)

* Risk ratio > 1 : People without patrent are more likely to churn > High risk = 1.22
* Risk ratio > 1 : less likely to churn > low risk = 0.75

We can't do the abve method for every varialbe
We can implement SQL in this , for example:
............................


SELECT
    gender,
    AVG(churn)
    AVG(churn) - global_churn AS diff,
    AVG(churn) - global_chrurn AS risk
FROM
    data
GROUP BY
    gender;


In [ ]:
df_full_train.groupby('gender').churn.mean() # To group the full training DataFrame by the 'gender' column and calculate the mean churn rate for each gender, providing insight into how gender may influence churn behavior in the dataset.

gender
female    0.276824
male      0.263214
Name: churn, dtype: float64

In [ ]:
df_group = df_full_train.groupby('gender').churn.agg(['mean', 'count']) # To group the full training DataFrame by the 'gender' column and calculate both the mean churn rate and the count of customers for each gender, providing a more comprehensive view of churn behavior and sample size for each demographic group.
df_group

,mean,count
gender,,
female,0.276824,2796
male,0.263214,2838


We can apply on all variables in categorical columns , we need to use spceal fnction to diaplay 

In [ ]:
from IPython.display import display # To import the display function from IPython.display, which allows for displaying rich content in Jupyter Notebooks, such as DataFrames and visualizations.

In [ ]:
for c in categorical: # To iterate over each column name in the list of categorical columns
    print(c) # To print the current column name being processed in the loop, providing context for the subsequent calculations and visualizations related to that specific categorical variable.
    
    df_group = df_full_train.groupby(c).churn.agg(['mean', 'count']) 
    
    df_group['diff'] = df_group['mean'] - global_churn_rate # To calculate the difference between the mean churn rate for each gender and the global churn rate, providing insight into how each gender's churn behavior deviates from the overall average churn rate in the dataset.
    df_group['risk'] = df_group['mean'] / global_churn_rate # To calculate the risk ratio for each gender by dividing the mean churn rate for each gender by the global churn rate, providing insight into how each gender's churn behavior compares to the overall average churn rate in the dataset.
    
    display(df_group) # To display the DataFrame 'df_group' in a rich format within the Jupyter Notebook, allowing for better visualization and interpretation of the calculated metrics for each gender.
    print()
    print()

gender


,mean,count,diff,risk
gender,,,,
female,0.276824,2796,0.006856,1.025396
male,0.263214,2838,-0.006755,0.974980




seniorcitizen


,mean,count,diff,risk
seniorcitizen,,,,
0,0.242270,4722,-0.027698,0.897403
1,0.413377,912,0.143409,1.531208




partner


,mean,count,diff,risk
partner,,,,
no,0.329809,2932,0.059841,1.221659
yes,0.205033,2702,-0.064935,0.759472




dependents


,mean,count,diff,risk
dependents,,,,
no,0.313760,3968,0.043792,1.162212
yes,0.165666,1666,-0.104302,0.613651




phoneservice


,mean,count,diff,risk
phoneservice,,,,
no,0.241316,547,-0.028652,0.893870
yes,0.273049,5087,0.003081,1.011412




multiplelines


,mean,count,diff,risk
multiplelines,,,,
no,0.257407,2700,-0.012561,0.953474
no_phone_service,0.241316,547,-0.028652,0.893870
yes,0.290742,2387,0.020773,1.076948




internetservice


,mean,count,diff,risk
internetservice,,,,
dsl,0.192347,1934,-0.077621,0.712482
fiber_optic,0.425171,2479,0.155203,1.574895
no,0.077805,1221,-0.192163,0.288201




onlinesecurity


,mean,count,diff,risk
onlinesecurity,,,,
no,0.420921,2801,0.150953,1.559152
no_internet_service,0.077805,1221,-0.192163,0.288201
yes,0.153226,1612,-0.116742,0.567570




onlinebackup


,mean,count,diff,risk
onlinebackup,,,,
no,0.404323,2498,0.134355,1.497672
no_internet_service,0.077805,1221,-0.192163,0.288201
yes,0.217232,1915,-0.052736,0.804660




deviceprotection


,mean,count,diff,risk
deviceprotection,,,,
no,0.395875,2473,0.125907,1.466379
no_internet_service,0.077805,1221,-0.192163,0.288201
yes,0.230412,1940,-0.039556,0.853480




techsupport


,mean,count,diff,risk
techsupport,,,,
no,0.418914,2781,0.148946,1.551717
no_internet_service,0.077805,1221,-0.192163,0.288201
yes,0.159926,1632,-0.110042,0.592390




streamingtv


,mean,count,diff,risk
streamingtv,,,,
no,0.342832,2246,0.072864,1.269897
no_internet_service,0.077805,1221,-0.192163,0.288201
yes,0.302723,2167,0.032755,1.121328




streamingmovies


,mean,count,diff,risk
streamingmovies,,,,
no,0.338906,2213,0.068938,1.255358
no_internet_service,0.077805,1221,-0.192163,0.288201
yes,0.307273,2200,0.037305,1.138182




contract


,mean,count,diff,risk
contract,,,,
month-to-month,0.431701,3104,0.161733,1.599082
one_year,0.120573,1186,-0.149395,0.446621
two_year,0.028274,1344,-0.241694,0.104730




paperlessbilling


,mean,count,diff,risk
paperlessbilling,,,,
no,0.172071,2313,-0.097897,0.637375
yes,0.338151,3321,0.068183,1.252560




paymentmethod


,mean,count,diff,risk
paymentmethod,,,,
bank_transfer_(automatic),0.168171,1219,-0.101797,0.622928
credit_card_(automatic),0.164339,1217,-0.105630,0.608733
electronic_check,0.455890,1893,0.185922,1.688682
mailed_check,0.193870,1305,-0.076098,0.718121
